core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -212,10 +212,9 @@ private void calculateCompactSize()
         boolean[] used = new boolean[dictionary.getPositionCount()];
         for (int i = 0; i < positionCount; i++) {
             int position = getId(i);
-            if (!used[position]) {
-                uniqueIds++;
-                used[position] = true;
-            }
+            // Avoid branch mispredictions by setting of position unconditionally.
```

Comment:
Are we sure this is because of branch mispredictions? Have you looked at the generated assembly to verify that the new form doesn't actually have a branch? Have you measured the actual number of branch mispredictions for the old code vs new code? You can do that with JMH on linux by enabling the LinuxPerfProfiler.